keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,960 @@
+
+### Shared files
+Certain hostVolume mounts are shared between kube-apiserver and kube-controller-manager, some examples of these are:-
+- /etc/kubernetes/pki/ca.crt
+- /etc/kubernetes/pki/sa.key
+
+We propose that files shared by kube-controller-manager and kube-apiserver be readable by a particular GID and we set the `supplementalGroups` in the [PodSecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritycontext-v1beta1-policy) of the Pod to this GID for kube-apiserver and kube-controller-manager.
+
+For instance lets consider that kubeadm grant group 2100 read permissions to /etc/kubernetes/pki/ca.crt
+
+Then we update the kube-apiserver manifest as follows:-
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    kubeadm.kubernetes.io/kube-apiserver.advertise-address.endpoint: 172.17.0.2:6443
+  creationTimestamp: null
+  labels:
+    component: kube-apiserver
+    tier: control-plane
+  name: kube-apiserver
+  namespace: kube-system
+spec:
+  securityContext:
+    supplementalGroups:
+    - 2100
+  containers:
+  - name: kube-apiserver
+    command:
+    - kube-apiserver
+    - --advertise-address=172.17.0.2
+    - --client-ca-file=/etc/kubernetes/pki/ca.crt
+    ... # omitted to save space 
+    securityContext:
+      runAsUser: 2000
+      runAsGroup: 2000
+      allowPrivilegeEscalation: false
+      seccompProfile:
+        type: runtime/default
+      capabilities:
+        drop:
+        - ALL
+    image: k8s.gcr.io/kube-apiserver:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    ... # omitted to save space
+    volumeMounts:
+    ... # omitted to save space
+    - mountPath: /etc/kubernetes/pki
+      name: k8s-certs
+      readOnly: true
+    ... # omitted to save space
+```
+
+Similarly kube-controller-manager's manifest would like:-
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: null
+  labels:
+    component: kube-controller-manager
+    tier: control-plane
+  name: kube-controller-manager
+  namespace: kube-system
+spec:
+  securityContext:
+    supplementalGroups:
+    - 2100
+  containers:
+  - name: kube-controller-manager
+    command:
+    - kube-controller-manager
+    - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
+    ...
+    # omitted to save space.
+    securityContext:
+      runAsUser: 2001
+      runAsGroup: 2001
+      allowPrivilegeEscalation: false
+      seccompProfile:
+        type: runtime/default
+      capabilities:
+        drop:
+        - ALL
+    image: k8s.gcr.io/kube-controller-manager:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    ... # omitted to save space.
+    volumeMounts:
+    - mountPath: /etc/kubernetes/pki
+      name: k8s-certs
+      readOnly: true
+    ... # omitted to save space. 
+```
+
+### Test Plan
```

Comment:
i'm interested if sig-security would like to see us create security tests on these features - e.g. making sure that user X cannot access the cluster CA private key....

Reply:
@vinayakankugoyal what tests do you envision here?

Reply:
I'll update this section once we have a some consensus on what approach we plan to take for assigning UID/GID.

Reply:
i think we need one regular test with feature-gate=true and one upgrade test with feature-gate=true

you can mention that we are going to create tests using this tooling: https://git.k8s.io/kubeadm/kinder
we need to double check if /etc/login.defs would work there as its nested container-runtimes (based on kind).

other than that, setting up the new tests will be fairly trivial.

Reply:
Updated, thanks!